File: src/com/owncloud/android/lib/common/http/methods/webdav/CopyMethod.java
package com.owncloud.android.lib.common.http.methods.webdav;

import java.net.URL;

import at.bitfire.dav4android.exception.UnauthorizedException;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import okhttp3.HttpUrl;
import okhttp3.Response;

public class CopyMethod extends DavMethod {

    final String destinationUrl;
    final boolean forceOverride;

    public CopyMethod(URL url, String destinationUrl, boolean forceOverride) {
        super(url);
        this.destinationUrl = destinationUrl;
        this.forceOverride = forceOverride;
    }

    @Override
    public int onExecute() throws Exception {
        try {
            mDavResource.copy(destinationUrl, forceOverride, response -> {
                mResponse = response;
                return Unit.INSTANCE;
            });

        } catch (UnauthorizedException davException) {
            // Do nothing, we will use the 401 code to handle the situation
        }
        return super.getStatusCode();
    }
}

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/CopyMethod.kt
/* ownCloud Android Library is available under MIT license
 *   Copyright (C) 2020 ownCloud GmbH.
 *
 *   Permission is hereby granted, free of charge, to any person obtaining a copy
 *   of this software and associated documentation files (the "Software"), to deal
 *   in the Software without restriction, including without limitation the rights
 *   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *   copies of the Software, and to permit persons to whom the Software is
 *   furnished to do so, subject to the following conditions:
 *
 *   The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 *
 *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 *   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 *   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 *   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 *   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 *   ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 *   CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *   THE SOFTWARE.
 *
 */
package com.owncloud.android.lib.common.http.methods.webdav

import at.bitfire.dav4jvm.DavOCResource
import okhttp3.Response
import java.net.URL

/**
 * Copy calls wrapper
 *
 * @author Christian Schabesberger
 * @author David González Verdugo
 */
class CopyMethod(
    val url: URL,
    private val destinationUrl: String,
    private val forceOverride: Boolean
) : DavMethod(url) {
    @Throws(Exception::class)
    public override fun onDavExecute(davResource: DavOCResource): Int {
        davResource.copy(
            destinationUrl,
            forceOverride,
            super.getRequestHeadersAsHashMap()
        ) { callBackResponse: Response ->
            response = callBackResponse
        }
        return super.statusCode
    }
}
